src/nupic/research/temporal_memory.py

```diff
@@ -340,7 +341,16 @@ def learnOnSegments(self,
     @param predictedInactiveCells       (set)         Indices of predicted inactive cells
     @param prevMatchingSegments         (set)         Indices of segments with
     """
-    for segment in prevActiveSegments | learningSegments:
+    segments = prevActiveSegments | learningSegments
+
+    # Sort segments before iterating for compatibility with C++
+    # Sort on secondary key (segment idx)
+    segments = sorted(segments)
+    # Sort on primary key (cell idx)
+    segments = sorted(segments,
```

Comment:
This is assuming a stable sort to preserve the secondary sort key from the previous sort? You should instead combine the sorts by having the lambda function return a tuple `(primaryKey, secondaryKey)`